type_ahead_dropdowns/main.js

```diff
@@ -0,0 +1,76 @@
+/* So 4 functions: filter drop down, show dropdown, update text input, and close drop down */
+
+;(function() {
+  // Displays markup on page
+  var renderMarkUp = function(array, parent) {
+    for (let item of array) {
+      parent.appendChild(item);
+    }
+    parent.classList.add('hidden');
+  };
+
+  // Generates Markup to be Displayed in renderMarkUp
+  var generateMarkUp = function(array) {
+    var options = data.map(function(el, index){
+      // for some really stupid reason I cant do option elements
+      var optionElement = document.createElement('input');
+      // because option.value don't work
+      optionElement.value = el;
+      optionElement.tabIndex = -index - 1;
+      optionElement.className = 'type-ahead__option';
+      optionElement.addEventListener('click', addOptionToInput);
+      return optionElement;
+    });
+    return options;
+  };
+
+  // Shows/Hides Markup
+  var toggleDropDown = function () {
+    markup.optionsContainer.classList.toggle('hidden');
+  };
+
+  // Filters MarkUp
+  var filterDropDownOptions = function() {
+    var criteria = event.target.value;
+    var items = document.querySelectorAll('.type-ahead__option');
+    for (let item of items) {
+      if (!item.value.includes(criteria)) {
+        item.classList.add('hidden');
+      } else if (item.value.includes(criteria)){
+        item.classList.contains('hidden') ? item.classList.remove('hidden') : null;
+      }
+    }
+    if (document.querySelectorAll('input.hidden').length === items.length) {
+     console.log('all inputs are hidden, whole container should be hidden');
+     markup.optionsContainer.classList.add('hidden');
+   } else if (document.querySelectorAll('input.hidden').length < items.length) {
+     console.log('NOT all inputs are hidden, container should be visible');
+     markup.optionsContainer.classList.contains('hidden') ?
+       markup.optionsContainer.classList.remove('hidden') :
+       null;
+   }
+  };
+
+  // Adds selected option to the textInput
+  var addOptionToInput = function() {
+    console.log('addOptionToInput is firing', event.target.value);
+    var selectedInput = event.target.value;
+    markup.dropDownInput.value = selectedInput;
+    toggleDropDown();
+  };
+
+  // I. Getting Data for the dropdown (to be retrived via HTTP Request)
+  var data = ['test', 'test2', 'test3', 'test4', 'test3','test3','test3','test3','test3','test3','test3'];
```

Comment:
Can you vary the strings in the data array? Makes it easier to see the filtering in action... (or the typeaheading)

Also consider that we might have an array of objects, and we might need to print a label, or a name.

Reply:
Commit 66b81a4 addresses this; type ahead now can accept either an array of objects or of strings; label element is now also included.